cmd/osquery-perf/agent.go

```diff
@@ -1607,33 +1660,37 @@ func scheduledQueryResults(packName, queryName string, numResults int) json.RawM
 }`)
 }
 
-func (a *agent) submitLogs(results []json.RawMessage) error {
-	// Connection check to prevent unnecessary JSON marshaling when the server is down.
-	conn, err := net.Dial("tcp", strings.TrimPrefix(a.serverAddress, "https://"))
+func (a *agent) connCheck() error {
+	request, err := http.NewRequest("GET", a.serverAddress+"/version", nil)
 	if err != nil {
-		return err
+		panic(err)
 	}
-	conn.Close()
-
-	jsonResults, err := json.Marshal(results)
+	response, err := http.DefaultClient.Do(request)
 	if err != nil {
-		panic(err)
+		return err
 	}
-	type submitLogsRequest struct {
-		NodeKey string          `json:"node_key"`
-		LogType string          `json:"log_type"`
-		Data    json.RawMessage `json:"data"`
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusOK {
+		return errors.New(http.StatusText(response.StatusCode))
 	}
-	slr := submitLogsRequest{
-		NodeKey: a.nodeKey,
-		LogType: "result",
-		Data:    jsonResults,
+	return nil
+}
+
+func (a *agent) submitLogs(results []resultLog) error {
+	// Connection check to prevent unnecessary JSON marshaling when the server is down.
+	if err := a.connCheck(); err != nil {
+		return fmt.Errorf("/version check failed: %w", err)
 	}
-	body, err := json.Marshal(slr)
-	if err != nil {
-		panic(err)
+
+	var resultLogs []byte
+	for i, result := range results {
+		if i > 0 {
+			resultLogs = append(resultLogs, ',')
+		}
+		resultLogs = append(resultLogs, result.emit()...)
 	}
 
+	body := []byte(`{"node_key": "` + a.nodeKey + `", "log_type": "result", "data": [` + string(resultLogs) + `]}`)
```

Comment:
Nit: Converting `resultLogs` from `[]byte` to `string` and back to `[]byte` body.

Reply:
Agree it could be optimized. I will merge as-is because making further changes will require another load test :)
(Otherwise there's the risk that I make a change and next developer running a load test will hit a bug.)